# Exploring the phylogenetic composition of microbiomes
### F. Mazel and J. Sanders

Welcome to the workshop !

## When and where? 
The workshop will be held at UBC (Vancouver) on the 20 of June 2018

Exact adress: Room 121 LSK; 6356 Agricultural road 

## Summary 

Microorganisms are the most widespread form of Life on Earth, inhabiting
most ecosystems (from animal guts to the deep ocean) and driving the bulk of global biogeochemical cycles. Yet, we still know little about the diversity distribution of microbiomes. Recent technical progresses, and in particular massive production of amplicon data --typically 16S reads-- have started to unravel microbiome diversity patterns. However, many questions remains. For example, what microbial unit should we use to study microbime compositional patterns (i.e. beta-diversity)? Should we use OTUs defined at 96% 16S similarity? ASV? In this workshop we propose to tackle this question by exploring the phylogenetic composition of microbiomes and thus avoiding a priori choices of microbial units. We will provide theoretical background and practical examples of such analysis using R.  


## To do BEFORE the workshop 

Please come to the workshop with your laptop set up with the required
software and data files as described in our [setup instructions](https://github.com/FloMazel/Microbiome_Phylo_Diversity_Workshop/blob/master/SetUp.Rmd).


## Content of this page

You will find 
1. A [document to help you set-up your laptop](https://github.com/FloMazel/Microbiome_Phylo_Diversity_Workshop/blob/master/SetUp.md)
2. A [document describing the workflow of the workshop](https://github.com/FloMazel/Microbiome_Phylo_Diversity_Workshop/blob/master/Workflow.Rmd)
3. The corresponding [R codes necessary to run the analysis](https://github.com/FloMazel/Microbiome_Phylo_Diversity_Workshop/blob/master/Main_R_script.R)


